lib/rucio/core/authentication.py

```diff
@@ -349,17 +446,856 @@ def get_auth_token_saml(account, saml_nameid, appid, ip=None, session=None):
+                                     refresh_lifetime=refresh_lifetime,
+                                     identity=validated['identity'],
+                                     ip=ip)
+            new_token.save(session=session)
+            session.expunge(new_token)
+            record_counter(counters='IdP_authorization.access_token.saved')
+            record_counter(counters='IdP_authorization.refresh_token.saved')
+            # remove refresh token info (not for the user/Rucio Client)
+            new_token.refresh_token = None
+            new_token.refresh_expired_at = None
+        else:
+            # save the received access token in the DB
+            new_token = models.Token(account=validated['account'],
+                                     token=oidc_tokens['access_token'],
+                                     scope=scope,
+                                     refresh=False,
+                                     expired_at=expired_at,
+                                     audience=audience,
+                                     identity=validated['identity'],
+                                     ip=ip)
+            new_token.save(session=session)
+            session.expunge(new_token)
+        record_counter(counters='IdP_authorization.exchange_token.saved')
+        record_timer(stat='IdP_authorization.exchange_token', time=time.time() - start)
+        return new_token
+
+    except:
+        return None
+
+
+@transactional_session
+def refresh_token_OIDC(token_object, session=None):
+    """
+    Requests new access and refresh tokens from the Identity Provider.
+    Assumption: The Identity Provider issues refresh tokens for one time use only and
+    with a limited lifetime. The refresh tokens are invalidated no matter which of these
+    situations happens first.
+
+    :param token_object: Rucio models.Token DB row object
+
+    :returns: None or throws an Exception.
+    """
+    global OIDC_clients
+    try:
+        start = time.time()
+        record_counter(counters='IdP_authorization.refresh_token.request')
+        refresh_start = datetime.datetime.utcnow()
+        # check if refresh token started in the past already
+        if hasattr(token_object, 'refresh_start'):
+            if token_object.refresh_start:
+                refresh_start = token_object.refresh_start
+        # check if refresh lifetime is set for the token
+        refresh_lifetime = 96
+        if token_object.refresh_lifetime:
+            refresh_lifetime = token_object.refresh_lifetime
+        # if the token has been refreshed for time exceeding the refresh_lifetime, the attempt will be aborted and refresh stopped
+        if (datetime.datetime.utcnow() - refresh_start > datetime.timedelta(hours=refresh_lifetime)):
+            # abort further refresh attempts
+            session.query(models.Token).filter(models.Token.token == token_object.token).update({models.Token.refresh: False})
+            session.commit()
```

Comment:
commit then raise? not rollback instead?

Reply:
I do commit, because I do want to change the DB row, normally I would log that this happened and exit the function, but as there is no logging I used the exception. I changed it so that the function returns a value in this case.